src/awkward/_attrs.py

```diff
@@ -41,4 +43,42 @@ def attrs_of(*arrays, attrs: Mapping | None = None) -> Mapping:
 
 
 def without_transient_attrs(attrs: dict[str, Any]) -> JSONMapping:
-    return {k: v for k, v in attrs.items() if not k.startswith("@")}
+    return {
+        k: v for k, v in attrs.items() if not (isinstance(k, str) and k.startswith("@"))
```

Comment:
I think we should simplify this and only allow strings. Attributes are nominally strong types, and by enforcing that we do not need to branch in every routine that handles attributes. Furthermore we don't then have to answer the question of what string-like objects should do (I.e non string subclasses).

We can possibly relax that down the road if there is a compelling case.

Reply:
I can enforce that all keys of `Attrs` are strings in the `Attrs` constructor. What do you think of this @jpivarski and @agoose77 ?
In principle I like it how it is now: users can store _anything_ in `.attrs` but not necessarily serialize it - pickle would complain properly in that case.

Reply:
I'm merging it as is, my thought process about this is that `.attrs` might well be used just for in-memory processing to keep track of _any_ metadata, it's not necessarily only used for serialization - so let's not restrict it preemptively for this (we would need to restrict values then aswell). If one pickles the array, `.attrs` obeys pickling rules and it will give reasonable errors to the user the same way as manually pickling python objects would. 
Also: it was possible to store anything until now, so restricting this might surprise some users? 

Within the awkward codebase we ensure using strings as keys through type annotations.

Hope this is fine for you, if not, I'm happy to address this in another PR :+1:

Reply:
@pfackeldey it was originally intended that attrs should only contain string keys -- e.g.
```
>>> import awkward
>>> import awkward as ak
>>> import pickle
>>> x = pickle.dumps(ak.Array([], attrs={1: 2}))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/angus/Git/awkward/src/awkward/highlevel.py", line 1644, in __reduce_ex__
    attrs = without_transient_attrs(self._attrs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/angus/Git/awkward/src/awkward/_attrs.py", line 44, in without_transient_attrs
    return {k: v for k, v in attrs.items() if not k.startswith("@")}
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/angus/Git/awkward/src/awkward/_attrs.py", line 44, in <dictcomp>
    return {k: v for k, v in attrs.items() if not k.startswith("@")}
                                                  ^^^^^^^^^^^^
AttributeError: 'int' object has no attribute 'startswith'
```

If we change that behaviour, we won't easily be able to undo it in future, where it will become breaking.

Reply:
Oh... I see what you mean with 'we restrict it already now for pickling', got you! Ok, I'll make sure we only store string keys - but only for the pickling step! I assume this is what you meant?

Reply:
To be super clear, the original intention of `attrs` is that it should hold string keys, and I think we should write our code as though that is true. I was replying to 
>  it's not necessarily only used for serialization - so let's not restrict it preemptively for this (we would need to restrict values then aswell).

Where I think we _should_ restrict preemptively given the above policy. If people really want to stash state in `attrs`, then they should have a custom namespace e.g. `attrs['my-namespace'] = {}`, I think!